### ISSUE TITLE
Forbid call to method Request::get()

### ISSUE THREAD

**Issue by vlakoff** [laravel/internals#114]
Laravel Request extends Symfony Request, thus it has a `get()` method. However, you are not expected to use it. Never. You should always use Laravel's `input()` instead. Otherwise you are likely to encounter some "hard to track down" bugs (immediately or even worse, later).

Refs:
- https://github.com/laravel/docs/pull/2269
- https://github.com/laravel/framework/issues/13138

So, maybe we should completely forbid its use, by extending it and throwing an exception.

**Comment by franzliedke** [comment]
But this breaks the Liskov Substitution Principle, right? 

In practice, this would mean that certain packages built for use with a Symfony request object would break when used with Laravel...

**Comment by vlakoff** [comment]
Indeed, still this wouldn't affect the Symfony class when used directly.

However, what could break is, if Laravel class calls a Symfony method, that in turn makes use of its `get()`. A solution may be to change the visibility, as so:

```php
protected function get() {
    return parent::get();
}
```

This way, we still allow calling this method from inside the classes, while removing it from the public API.

**Comment by jlaswell** [comment]
Since Laravel's Request object extends Symfony's Request, the `get()` method must [stay public](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/Request.php#L702-L731). Even if it could be switched to protected, the principle is still broken since the usage is restricted to inheriting classes.

**Comment by vlakoff** [comment]
I still think there should be some way to prevent users from doing `$request->get()`. It looks so innocent, but it is unexpected and introduces subtle bugs.

The [visibility change](https://github.com/laravel/ideas/issues/114#issuecomment-225946972) I suggested above should work well in practice. But maybe someone has a better idea?

**Comment by mfn** [comment]
Code changes which throw exceptions or change the visibility in this case are a big no-no.

Education is the best weapon here, but mentioning in the docs something you ought not to use is weird either (as in "bringing something into focus for the purpose of actually not having it in focus").

Another suggestion: override the method, only call it's parent but add a phpdoc explaining the problem with it.

**Comment by vlakoff** [comment]
This sounds great to me.

**Comment by sisve** [comment]
Override the method and forward the call to ´$this->input(...)` instead? We could at least make it work The Laravel Way.

**Comment by vlakoff** [comment]
The commenters above would say it's a big no-no, for the same reasons.

And here I would agree with them, it would monstrously break the Liskov substitution principle, and certainly many user codes.

**Comment by vlakoff** [comment]
Suggestions for phrasing the phpdoc are welcome. Also, should we put the `@param`'s and `@return`?

If it may help, here's a link to the [Symfony method](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/Request.php#L685).

**Comment by sisve** [comment]
How would forwarding `get($key)` to `input($key)` break LSP? A LaravelRequest can still be sent to a method expecting a SymfonyRequest, and someone calling ->get() on it will still get the data, just using Laravel's rules of order and not the Symfony order.

Forwarding the method call makes sense for security purposes, the validation has read data using the same logic as input(), and get() can read invalid data after the validator has said that the request is valid.

**Comment by vlakoff** [comment]
Maybe the bare signature would be compatible, but it would change the behaviour, which for me is a clear break of the LSB (emphasis mine):

> ... objects of type T in a program may be replaced with objects of type S **without altering any of the desirable properties of that program** (e.g. [correctness](https://en.wikipedia.org/wiki/Correctness_(computer_science))).

**Comment by sisve** [comment]
I still don't see it. Code calls `->get($key)`, and they get the value they asked for. A Laravel request is still a Symfony request, and can still be used as such.

The entire paragraph is relevant for context; it shows that LSP is about classes and not small changes to invididual methods.

> Liskov's notion of a behavioral subtype defines a notion of substitutability for objects; that is, if S is a subtype of T, then objects of type T in a program may be replaced with objects of type S without altering any of the desirable properties of that program (e.g. correctness).

The [wikipedia article](https://en.wikipedia.org/wiki/Liskov_substitution_principle) then continues with a list of things that LSP is about. The first part is about types as type safety, and not relevant to this case. This is usually enforced by a compiler in strongly typed languages.

> Liskov's principle imposes some standard requirements on signatures that have been adopted in newer object-oriented programming languages (usually at the level of classes rather than types; see nominal vs. structural subtyping for the distinction):
> * Contravariance of method arguments in the subtype.
> * Covariance of return types in the subtype.
> * No new exceptions should be thrown by methods of the subtype, except where those exceptions are themselves subtypes of exceptions thrown by the methods of the supertype.

These three points, which are still not relevant to this issue, means that...
1. If the base method accepts a DOMElement, then the overriding method can accept a parent class of DOMElement, like a DOMNode. This works because everyone that knows about the base class knows it supports the DOMElement, and you've just extended the support to also allow DOMNode objects while still keeping the support for DOMElement.
2. If the base method returns a DOMNode, then your overriding method can return a class that derives from DOMNode, like a DOMElement. This works because everyone knows that base class will return a node, your code just happens to always return an element which is still a node.
3. This is about people that has try/catch statements, and not introducing new exceptions that aren't caught by the existing catch-clauses.

> In addition to the signature requirements, the subtype must meet a number of behavioral conditions. These are detailed in a terminology resembling that of design by contract methodology, leading to some restrictions on how contracts can interact with inheritance:
> * Preconditions cannot be strengthened in a subtype.
> * Postconditions cannot be weakened in a subtype.
> * Invariants of the supertype must be preserved in a subtype.
> *  History constraint (the "history rule"). Objects are regarded as being modifiable only through their methods (encapsulation). Because subtypes may introduce methods that are not present in the supertype, the introduction of these methods may allow state changes in the subtype that are not permissible in the supertype. The history constraint prohibits this. It was the novel element introduced by Liskov and Wing. A violation of this constraint can be exemplified by defining a mutable point as a subtype of an immutable point. This is a violation of the history constraint, because in the history of the immutable point, the state is always the same after creation, so it cannot include the history of a mutable point in general. Fields added to the subtype may however be safely modified because they are not observable through the supertype methods. Thus, one can derive a circle with fixed center but mutable radius from immutable point without violating LSP.

1. If the base method had an constraint/assertion that `$account->cash >= 0`, then your code is expected to work with this and not suddenly required `$account->cash == 1337`. Your code should work within the same premises as the original code.
2. If the base method had a constraint/assertion that `$account->loggedIn == true` to only return logged in users, then your code cannot remove this to return all users. Your new code's behavior would no longer match, and people using the old code to list logged in users would no longer have that behavior.
3. If the base method makes a guarantee of something, then your derived class must make the same guarantee. For example, the base method says "hey, I support all encodings", then you're having a bad day and your code needs to support those too.
4. This is a long version of saying "your new method should not modify the base class members in a way that the base class does not permit itself".

Why are we even arguing about LSP here? We're doing weirdness and breaking changes all over the framework all the time. Why would we not change an existing method into returning proper validated data?

TL;DR: So, I see nothing in the type safety rules or behavior rules of LSP that would stop us from forwarding `get($key)` to `input($key)`.

**Comment by franzliedke** [comment]
If you construct an `Illuminate\Http\Request` object `$a` and a `Symfony\HttpFoundation\Request` object `$b` to represent the exact same HTTP request, then `$a->get($foo)` should **always** return the same data as `$b->get($foo)`, no matter the request we're dealing with.

That's how I understand LSP.

Otherwise, you'd sooner or later run into a situation where a package expecting to deal with Symfony request objects would break down when dealing with an Illuminate request object.

(To be fair, I haven't delved into the details of how `input()` and `get()` work, but from a quick glance, the implementations look different enough that constructing such a scenario should be possible.)

**Comment by nCrazed** [comment]
> If you construct an Illuminate\Http\Request object $a and a Symfony\HttpFoundation\Request object $b to represent the exact same HTTP request, then $a->get($foo) should always return the same data as $b->get($foo), no matter the request we're dealing with.

If $b->get() can never return different data, we can no longer use inheritance for polymorphism.

**Comment by franzliedke** [comment]
Okay, that might have been a little too broad.

Unfortunately, "desirable properties" is a very imprecise term. :wink: 

Let me phrase it this way: if I can construct a situation, where an `Illuminate\Http\Request` object can be created that fails any of the tests defined for Symfony's request class, then that should be considered as breaking LSP. If anything comes close to define the "desirable properties", it would be test cases...

**Comment by nCrazed** [comment]
The point still stands. Polymorphic subclasses will fail some of such tests, because they behave  differently by design.

**Comment by franzliedke** [comment]
If that doesn't contradict LSP, then I don't know what does. :wink:

**Comment by nCrazed** [comment]
Which part of LSP is violated in the following example?

You have a 3rd-party `DateSource` class that exposes `function get():array` which in the default implementation will return all key-value pairs returned from the data source (database, api, whatever.). You also have some 3rd-party consumers that happen to accept `DataSource`.

If you want to hide some of the sensitive key-value pairs from such consumers, you implement a `PublicDataSource` that performs some filtering in its implementation of `function get(): array`.

Your `PublicDataSource` will correctly fail some of the tests written specifically for the base `DataSource` class. If it didn't, it would have no reason to exist. It will also be valid and accepted anywhere a `DataSource` is accepted. As far as any consumer is concerned, the returned pairs are all pairs that are available.

**Comment by vlakoff** [comment]
I think it would precisely be a misuse of inheritance. In this example, `PublicDataSource` shouldn't extend `DataSource` and you should rather use [composition](https://en.wikipedia.org/wiki/Object_composition):

```php
class PublicDataSource {

    protected $dataSource; // instance of DataSource

    public function get($key) {

        // just a basic example
        if (Str::startsWith($key, 'hidden_')) {
            return false;
        }

        return $this->dataSource->get($key);
    }
}
```

Long story short, unsubscribing. Couldn't we just apply [mfn's suggestion](https://github.com/laravel/ideas/issues/114#issuecomment-373509373)...

**Comment by shalvah** [comment]
This was an entertaining discussion. 😄
